services/scan/src/plustek_machine.ts

```diff
@@ -0,0 +1,809 @@
+    case PaperStatus.VtmFrontAndBackSensorHavePaperReady:
+      return { type: 'SCANNER_JAM' };
+    default:
+      throw new Error(`Unexpected paper status: ${paperStatus}`);
+  }
+}
+
+function paperStatusErrorToEvent(
+  error: ScannerError | Error
+): ScannerStatusEvent {
+  if (error instanceof ClientDisconnectedError) {
+    return { type: 'SCANNER_DISCONNECTED' };
+  }
+  switch (error) {
+    // This error code happens when the paper gets blocked and tries to scan for
+    // a long time (or if you just hold onto the paper while scanning).
+    case ScannerError.SaneStatusInval:
+      return { type: 'SCANNER_JAM' };
+    case ScannerError.SaneStatusIoError:
+      return { type: 'SCANNER_DISCONNECTED' };
+    default:
+      throw error;
+  }
+}
+
+// Create a paper status observable, then use internal transitions to avoid
+// changing state when paper status doesn't change
+function paperStatusObserver({ client }: Context) {
+  assert(client);
+  return timer(0, 500).pipe(
+    switchMap(async () => {
+      const paperStatus = await client.getPaperStatus();
+      debugPaperStatus('Paper status: %s', paperStatus);
+      return paperStatus.isOk()
+        ? paperStatusToEvent(paperStatus.ok())
+        : paperStatusErrorToEvent(paperStatus.err());
+    })
+  );
+}
+
+const pollPaperStatus: InvokeConfig<Context, Event> = {
+  src: paperStatusObserver,
+  onError: {
+    target: 'error',
+    actions: assign({ error: (_, event) => event.data }),
+  },
+};
+
+async function scan({ client }: Context): Promise<SheetOf<string>> {
+  assert(client);
+  debug('Scanning');
+  const scanResult = await client.scan();
+  debug('Scan result: %o', scanResult);
+  const [front, back] = scanResult.unsafeUnwrap().files;
+  return [front, back];
+}
+
+async function interpretSheet({
+  interpreter,
+  scannedSheet,
+  interpretationMode,
+}: Context): Promise<InterpretationResultEvent> {
+  assert(interpreter);
+  assert(scannedSheet);
+
+  if (interpretationMode === 'skip') {
+    return {
+      type: 'INTERPRETATION_VALID',
+      interpretation: {
+        sheetId: 'mock-sheet-id',
+        sheet: [
+          {
+            originalFilename: '/front-original-mock',
+            normalizedFilename: '/front-normalized-mock',
+            interpretation: { type: 'BlankPage' },
+          },
+          {
+            originalFilename: '/back-original-mock',
+            normalizedFilename: '/back-normalized-mock',
+            interpretation: { type: 'BlankPage' },
+          },
+        ],
+      },
+    };
+  }
+
+  const sheetId = uuid();
+  const result = await interpreter.interpret(sheetId, scannedSheet);
+  const interpretedSheet = result.unsafeUnwrap();
+  const [front, back] = interpretedSheet;
+  const frontType = front.interpretation.type;
+  const backType = back.interpretation.type;
+  const interpretation: Interpretation = { sheetId, sheet: interpretedSheet };
+
+  if (frontType === 'BlankPage' && backType === 'BlankPage') {
```

Comment:
We should try to consolidate this stuff, but it might take a lot of effort to do so properly. IMO we should not have separate `PageInterpretation` objects for each page. Instead, we should have a `SheetInterpretation` that combines them together, marks/votes/errors etc.

Also, this particular logic looks wrong. A `BlankPage` means a literally white page with essentially nothing on it. A `BlankBallot` is a ballot without any marks. Confusing, I know 😢

Reply:
Ok yeah that makes sense, I like that vision. Might expand my scope a bit too much - we'll see.

Good callout on the `BlankPage` - thanks. Seems like `BlankPage` on both sides isn't actually a real case that could happen then?

Reply:
It'll happen if someone puts in a fully blank sheet, but that's it.